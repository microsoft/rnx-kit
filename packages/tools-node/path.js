module.exports = require("./lib/path");
