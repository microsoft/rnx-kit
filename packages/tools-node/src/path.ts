/**
 * Escape a path by replacing each backslash ('\\') with a double-backslash ("\\\\").
 *
 * @param p Path to escape
 * @returns Escaped path
 */
export function escapePath(p: string): string {
  return p.replace(/\\/g, "\\\\");
}

/**
 * Normalize the separators in a path, converting each backslash ('\\') to a foreward
 * slash ('/').
 *
 * @param p Path to normalize
 * @returns Normalized path
 */
export function normalizePath(p: string): string {
  return p.replace(/\\/g, "/");
}
