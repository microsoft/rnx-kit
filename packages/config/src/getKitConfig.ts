import path from "path";
import { readPackage } from "@rnx-kit/tools-node";
import { KitConfig } from "./kitConfig";

/**
 * Options for retrieving a kit config. The default is equivalent to passing { cwd: process.cwd() }
 */
export type GetKitConfigOptions = {
  /**
   * Retrieve the rnx-kit configuration from a package module.
   */
  module?: string;

  /**
   * Retrive the rnx-kit configuration from a directory containing 'package.json'.
   */
  cwd?: string;
};

/**
 * Query for a package's rnx-kit configuration.
 *
 * @param options Options for retrieving the configuration.
 * @returns
 */
export function getKitConfig(
  options: GetKitConfigOptions = {}
): KitConfig | undefined {
  // find the package dir that holds the rnx-kit configuration
  const packageDir = options.module
    ? path.dirname(require.resolve(options.module + "/package.json"))
    : options.cwd || process.cwd();

  // try to read package.json
  try {
    const packageJson = readPackage(packageDir);
    return packageJson["rnx-kit"];
  } catch {
    return undefined;
  }
}
