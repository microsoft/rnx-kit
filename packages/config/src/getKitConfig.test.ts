// @ts-check
"use strict";

import { getKitConfig } from "./getKitConfig";
import path from "path";

describe("@rnx-kit/config tests", () => {
  const notPresentFixture = "not-present";
  const inPackageFixture = "in-package";
  const inConfigFixture = "in-config-js";

  const currentWorkingDir = process.cwd();

  function fixturePath(name: string): string {
    return path.join(currentWorkingDir, "src", "__fixtures__", name);
  }

  /**
   * Sets current working directory to specified test fixture.
   * @param {string} name
   */
  function setFixture(name: string): void {
    process.chdir(fixturePath(name));
  }

  afterEach(() => process.chdir(currentWorkingDir));

  test("not-present package correctly returns a null config", () => {
    setFixture(notPresentFixture);
    expect(getKitConfig()).toBeNull();
  });

  test("kit options can be found in package.json", () => {
    setFixture(inPackageFixture);
    const config = getKitConfig();
    expect(config).not.toBeNull();
    expect(config && config.bundle).toBeTruthy();
    expect(config && config.platformBundle).toBeTruthy();
    expect(config).toMatchSnapshot();
  });

  test("kit options can be found in config file", () => {
    setFixture(inConfigFixture);
    const config = getKitConfig();
    expect(config).not.toBeNull();
    expect(config).toMatchSnapshot();
  });

  test("kit options can be loaded via path", () => {
    const config = getKitConfig({ cwd: fixturePath(inPackageFixture) });
    expect(config).toMatchSnapshot();
  });
});
