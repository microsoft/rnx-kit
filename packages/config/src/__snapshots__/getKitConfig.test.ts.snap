// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@rnx-kit/config tests kit options can be found in config file 1`] = `
Object {
  "bundle": Object {
    "assetsPath": "./build-out/assets",
    "bundlePrefix": "main",
    "distPath": "./build-out",
    "entryPath": "./core-entry.js",
    "id": "core",
    "platforms": Object {
      "android": Object {
        "assetsPath": "./build-out/res",
      },
    },
    "targets": Array [
      "ios",
      "android",
      "macos",
      "windows",
    ],
  },
  "dependencies": [Function],
  "platformBundle": Object {
    "bundlePrefix": "fakeIndex",
    "distPath": "./distro",
  },
}
`;

exports[`@rnx-kit/config tests kit options can be found in package.json 1`] = `
Object {
  "bundle": true,
  "dependencies": Object {
    "bar": "2.1.1",
    "foo": "0.1.1",
  },
  "platformBundle": true,
}
`;

exports[`@rnx-kit/config tests kit options can be loaded via path 1`] = `
Object {
  "bundle": true,
  "dependencies": Object {
    "bar": "2.1.1",
    "foo": "0.1.1",
  },
  "platformBundle": true,
}
`;
